infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,60 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once upgrades are done it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌────────────────────────────────┐
+ *                  │ isClusterAtV2 = false (in-mem) │
+ *                  └──────────────────┬─────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼───────────┐                         ┌──────▼───────────────┐
+ * │ isClusterAtV2 = false │                         │ isClusterAtV2 = true │
+ * └───────────┬───────────┘                         └───────┬──────────────┘
+ *             │            ┌──────────────────────┐         │
+ *             └───────────►│ start Log Replicator │◄────────┘
+ *                          └──────────────────────┘
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isClusterAtV2 = false;
+    ILogReplicationVersionAdapter versionAdapter;
+    public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+        if (isClusterAtV2) {
+            return false;
+        }
+        boolean isClusterUpgradeInProgress = versionAdapter.isUpgradeRunning(txnContext);
+        if (isClusterUpgradeInProgress) {
+            return true;
+        } // else implies cluster upgrade has completed
+        migrateData(txnContext);
+        isClusterAtV2 = true;
+        return false;
+    }
```

Comment:
Sorry I missed the initial discussion of this change, I'm a bit confused by why the `migrateData()` is specific to LR? It looks like the data migration is something to be performed at Corfu cluster level after rolling upgrade is done? i.e. read state from old format and write in new format, delete old formats, etc.

Reply:
Got it now, thanks!